### ISSUE TITLE
Images that have been processed with histogram equalization don't decode/encode properly, but images that have been processed with other functions and decoded/encoded are fine.

### ISSUE THREAD

**Issue by kofferdahl** [matthew-huber/bme590final#73]
For a demo of this error, checkout the branch fixHistogramEqualizationBug, and look at the file debugging_histogram_equalization.py. 
https://github.com/matthew-huber/bme590final/blob/fixHistogramEqualizationBug/debugging_histogram_equalization.py

This file is currently configured to demonstrate that an image processed with histogram equalization, and then encoded / decoded results in a completely black image. However, that image that was processed with histogram equalization before being encoded / decoded looks correct - you can verify this for yourself by uncommenting lines 14 and 15 in debugging_histogram_equalization.py
https://github.com/matthew-huber/bme590final/blob/fcaba53ea1f1b9514eb9c638e2eec5624c32be19/debugging_histogram_equalization.py#L14

This error does not occur when the same image is processed using a different processing type. To verify this, uncomment line 11 to perform contrast stretching instead of histogramEqualization. Here, you should show that the image looks as expected after being encoded/decoded. 
https://github.com/matthew-huber/bme590final/blob/fcaba53ea1f1b9514eb9c638e2eec5624c32be19/debugging_histogram_equalization.py#L11

We aren't sure why this error would uniquely occur only when images are processed with histogram equalization and decoded/encoded, considering that the other images encode/decode properly, and the histogram equalized image that hasn't been encoded/decoded is also fine.

**Comment by mlp6** [comment]
Bigger question first... you have your `ImageProcessor` class, but why not have `img` be an attribute of that class that then the methods can act on?  I ask b/c it would be nice to couple the `img` datatype with the methods that will work on it.

**Comment by mlp6** [comment]
To your specific question, have you checked the resultant type of data returned by each of the methods?

**Comment by mjp59** [comment]
fixed it. 
Turns out doing a histogram equalization makes all the values in the processed image array between 0 and 1. So the reformatting to a readable image file was rounding these values off to 0 or 1. 

The values of the image should be discrete values from 0 to 255. To make this true, take the processed image and multiply by 255 so 1 maps to 255 and 0 maps to 0 giving 256 discrete values. This gives the correct image that has undergone histogram equalization.

**Comment by mjp59** [comment]
@kofferdahl There seems be a bug that is occurring in all the functions. Where the image processor methods are being applied correctly but then the red and blue channels values are being swapped for color images, when they should not be swapped. This bug seems to be causing the difference. So I think histogram equal is fine, but there a bug that affects all methods rn.

**Comment by kofferdahl** [comment]
@mjp59 I think I found at least one problem. When the image is rescaled to 0 to 255, the numbers are still floats rather than ints. I think this causes some kind of rounding / issues when the image is displayed. I've got a new version of histogram_equalization that performs on each color channel individually, rescales it, and then converts from float64 to uint8. This at least makes the histogram equalization images look better than they did before. They still don't quite look right, but maybe after the channel swapping problem is fixed they'll be okay.
